># **Pertemuan 12 - Ensemble Model (Random Forest)**

In [19]:
import pandas as pd
import numpy as np

<small> 1. Gunakanlah dataset “ChurnDataset.csv”. </small>

In [20]:
data = pd.read_csv('ChurnDataset.csv')

<small> 2.  Lakukanlah data preparations. Perhatikan kolom-kolom yang bertipe data kategorikal, anda
dapat menggunakan one-hot encoding/ordinal encoding/label encoder. </small>

In [21]:
from sklearn.preprocessing import OneHotEncoder, LabelEncoder

print(data.head())
data = data.drop(columns=\
['RowNumber', 'CustomerId', 'Surname'])

data = pd.get_dummies(data, \
columns=['Geography'], drop_first=True)

label_encoder = LabelEncoder()
data['Gender'] = label_encoder.fit_transform(data['Gender'])

X = data.drop(columns=['Exited'])
y = data['Exited']

   RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0       2       0.00              1          1               1   
1       1   83807.86              1          0               1   
2       8  159660.80              3          1               0   
3       1       0.00              2          0               0   
4       2  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         790

<small> 3. Pisahkan dataset yang ada menjadi data train dan data test dengan rasio 70% dan 30% </small>

In [4]:
from sklearn.model_selection import train_test_split, GridSearchCV

In [5]:
X_train, X_test, y_train, y_test = \
train_test_split(X, y, test_size=0.3, random_state=42)

<small> 4.Lakukan pembentukan base model dengan menggunakan SVM, Decision Tree dan Random
Forest. </small>

In [6]:
from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report, confusion_matrix

In [11]:
# Membangun model SVM
svm_model = SVC(random_state=42)
svm_model.fit(X_train, y_train)

SVC(random_state=42)

In [12]:
# Membangun model Decision Tree
decision_tree_model = DecisionTreeClassifier\
(random_state=42)
decision_tree_model.fit(X_train, y_train)

DecisionTreeClassifier(random_state=42)

In [13]:
# Membangun model Random Forest
random_forest_model = RandomForestClassifier\
(random_state=42)
random_forest_model.fit(X_train, y_train)

RandomForestClassifier(random_state=42)

<small> 5.Lakukanlah pembentukan Grid Search untuk mencari parameter terbaik dengan cross-validation (minimal 2 parameter). </small>

In [ ]:
# Parameter grid untuk SVM
svm_param_grid = {
    'C': [0.1, 1, 10],
    'kernel': ['linear', 'rbf']
}

svm_grid_search = GridSearchCV\
(svm_model, svm_param_grid, \
cv=5, n_jobs=-1, verbose=1)

svm_grid_search.fit(X_train, y_train)
print("Best parameters for SVM:", svm_grid_search.best_params_)

Fitting 5 folds for each of 6 candidates, totalling 30 fits


In [16]:
# Parameter grid untuk Decision Tree
decision_tree_param_grid = {
    'max_depth': [None, 10, 20, 30],
    'min_samples_split': [2, 10, 20]
}

decision_tree_grid_search = GridSearchCV\
(decision_tree_model, decision_tree_param_grid,\
cv=5, n_jobs=-1, verbose=1)

decision_tree_grid_search.fit(X_train, y_train)

print("Best parameters for Decision Tree:",\
decision_tree_grid_search.best_params_)

Fitting 5 folds for each of 12 candidates, totalling 60 fits
Best parameters for Decision Tree: {'max_depth': 10, 'min_samples_split': 20}


In [17]:
# Parameter grid untuk Random Forest
random_forest_param_grid = {
    'n_estimators': [100, 200, 300],
    'max_depth': [None, 10, 20],
}

random_forest_grid_search = GridSearchCV(random_forest_model\
, random_forest_param_grid, cv=5, n_jobs=-1, verbose=1)

random_forest_grid_search.fit(X_train, y_train)
print("Best parameters for Random Forest:",\
random_forest_grid_search.best_params_)

Fitting 5 folds for each of 9 candidates, totalling 45 fits
Best parameters for Random Forest: {'max_depth': None, 'n_estimators': 300}


<small>6.Lakukan pembentukan model menggunakan SVM, Decision Tree dan Random Forest dengan
best parameter yang diambil dari hasil grid search dan tampilkan hasil akurasi yang didapat.</small>

In [33]:
from sklearn.metrics import accuracy_score, \
f1_score, confusion_matrix, \
precision_score, recall_score

In [28]:
# Membentuk model SVM dengan best parameters
best_svm_params = svm_grid_search.best_params_
best_svm_model = SVC(**best_svm_params, \
random_state=42)

best_svm_model.fit(X_train, y_train)
best_svm_predictions = best_svm_model.predict(X_test)
best_svm_accuracy = accuracy_score(y_test, best_svm_predictions)
print("SVM Best Accuracy:", best_svm_accuracy)

AttributeError: 'GridSearchCV' object has no attribute 'best_params_'

In [26]:
# Membentuk model Decision Tree dengan best parameters
best_decision_tree_params = decision_tree_grid_search.best_params_
best_decision_tree_model = \
DecisionTreeClassifier(**best_decision_tree_params, random_state=42)

best_decision_tree_model.fit(X_train, y_train)
best_decision_tree_predictions = \
best_decision_tree_model.predict(X_test)

best_decision_tree_accuracy = \
accuracy_score(y_test, best_decision_tree_predictions)

print("Decision Tree Best Accuracy:", best_decision_tree_accuracy)

Decision Tree Best Accuracy: 0.839


In [27]:
# Membentuk model Random Forest dengan best parameters
best_random_forest_params = random_forest_grid_search.best_params_
best_random_forest_model = \
RandomForestClassifier(**best_random_forest_params, random_state=42)

best_random_forest_model.fit(X_train, y_train)
best_random_forest_predictions = \
best_random_forest_model.predict(X_test)

best_random_forest_accuracy = \
accuracy_score(y_test, best_random_forest_predictions)
print("Random Forest Best Accuracy:", best_random_forest_accuracy)

Random Forest Best Accuracy: 0.8666666666666667


<small>7.Tampilkan accuracy, confusion matrix(F1 Score), precision serta recall dari kedua model anda.</small>

In [37]:
# Accuracy, confusion matrix, precision, recall SVM
svm_pred = svm_model.predict(X_test)
print(f"SVM Accuracy: \
{accuracy_score(y_test, svm_pred)}")

print(f"SVM Confusion Matrix\
:{confusion_matrix(y_test, svm_pred)}")

print(f"SVM Precision: \
{precision_score(y_test, svm_pred)}")

print(f"SVM Recall:\
{recall_score(y_test, svm_pred)}")

# Untuk model dengan parameter terbaik SVM
print(f'SVM Best Accuracy:\
{best_svm_accuracy}')

print(f'Best Confusion Matrix:\
{confusion_matrix(y_test, best_svm_predictions)}')

print(f'Best SVM F1 Score: \
{f1_score(y_test,best_svm_predictions)}')

print(f'Best SVM Precision: \
{precision_score(y_test,best_svm_predictions)}')

print(f'Best SVM Recall: \
{y_test,best_svm_predictions}')

SVM Accuracy: 0.8053333333333333
SVM Confusion Matrix:[[2416    0]
 [ 584    0]]
SVM Precision: 0.0
SVM Recall: 0.0


/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


NameError: name 'best_svm_accuracy' is not defined

In [40]:
# Accuracy, confusion matrix, precision, recall Decision Tree
decision_tree_pred = decision_tree_model.predict(X_test)
print(f"Decision Tree Accuracy: \
{accuracy_score(y_test, decision_tree_pred)}")

print(f"Decision Tree Confusion Matrix:\
{confusion_matrix(y_test, decision_tree_pred)}")

print(f"Decision Tree Precision: \
{precision_score(y_test, decision_tree_pred)}")

print(f"Decision Tree Recall: \
{recall_score(y_test, decision_tree_pred)}\n")


# Untuk model dengan parameter terbaik Decision Tree
print(f"Decision Tree Best Accuracy: \
{best_decision_tree_accuracy}")

print(f'Best Confusion Matrix: \
{confusion_matrix(y_test, best_decision_tree_predictions)}')

print(f'Best Decision Tree F1 Score: \
{f1_score(y_test, best_decision_tree_predictions)}')

print(f'Best Decision Tree Precision: \
{precision_score(y_test, best_decision_tree_predictions)}')

print(f'Best Decision Tree Recall: \
{recall_score(y_test, best_decision_tree_predictions)}')

Decision Tree Accuracy: 0.808
Decision Tree Confusion Matrix:[[2113  303]
 [ 273  311]]
Decision Tree Precision: 0.506514657980456
Decision Tree Recall: 0.5325342465753424

Decision Tree Best Accuracy: 0.839
Best Confusion Matrix: [[2247  169]
 [ 314  270]]
Best Decision Tree F1 Score: 0.527859237536657
Best Decision Tree Precision: 0.6150341685649203
Best Decision Tree Recall: 0.4623287671232877


In [41]:
# Accuracy, confusion matrix, precision, recall Random Forest
random_forest_pred = random_forest_model.predict(X_test)
print(f"Random Forest Accuracy: \
{accuracy_score(y_test, random_forest_pred)}")

print(f"Random Forest Confusion Matrix:\
{confusion_matrix(y_test, random_forest_pred)}")

print(f"Random Forest Precision:\
{precision_score(y_test, random_forest_pred)}")

print(f"Random Forest Recall:\
{recall_score(y_test, random_forest_pred)}\n")

# Untuk model dengan parameter terbaik Random Forest
print(f"Random Forest  Best Accuracy: \
{best_random_forest_accuracy}")

print(f'Best Confusion Matrix: \
{confusion_matrix(y_test, best_random_forest_predictions)}')

print(f'Best Random Forest F1 Score: \
{f1_score(y_test, best_random_forest_predictions)}')

print(f'Best Random Forest Precision: \
{precision_score(y_test, best_random_forest_predictions)}')

print(f'Best Random Forest Recall: \
{recall_score(y_test, best_random_forest_predictions)}')

Random Forest Accuracy: 0.867
Random Forest Confusion Matrix:[[2328   88]
 [ 311  273]]
Random Forest Precision:0.7562326869806094
Random Forest Recall:0.4674657534246575

Random Forest  Best Accuracy: 0.8666666666666667
Best Confusion Matrix: [[2330   86]
 [ 314  270]]
Best Random Forest F1 Score: 0.5744680851063829
Best Random Forest Precision: 0.7584269662921348
Best Random Forest Recall: 0.4623287671232877


<small>8.Berikan analisis apakah hasilnya menjadi lebih baik</small>

Untuk SVM :
(belum selesai running karena lama sekali)

Untuk Decision Tree:

Setelah penyetelan hyperparameter, akurasi, presisi, dan recall meningkat. Akurasi meningkat dari 0.808 menjadi 0.839, presisi meningkat dari 0.506 menjadi 0.615, namun recall mengalami penurunan dari 0.533 menjadi 0.462.

Untuk Random Forest:
Setelah penyetelan hyperparameter, akurasi dan presisi Random Forest tetap relatif stabil, namun recall mengalami sedikit penurunan dari 0.467 menjadi 0.462.

Secara keseluruhan, meskipun terjadi perbaikan pada beberapa metrik kinerja model Decision Tree setelah penyetelan hyperparameter, peningkatan ini tidak signifikan karena recall mengalami penurunan. Sementara itu, performa model Random Forest cenderung stabil setelah penyetelan hyperparameter, dengan sedikit perubahan pada recall. Hal ini menunjukkan bahwa penyetelan hyperparameter tidak selalu menghasilkan peningkatan kinerja yang signifikan pada setiap model.